### ISSUE TITLE
Extract better python semantic embeddings for similarity comparison 

### ISSUE THREAD

**Issue by lazyhope** [microsoft/CodeBERT#217]
I want to compare python files from two repositories, the cross-encoder way of clone detection doesn't scale well for this goal, therefore I want to encode each code into embeddings and compare them instead.

I realized UnixCoder generates better code embeddings than GraphCodeBERT in this case, but I am wondering if fine-tuning it for one of it's downstream tasks makes it produce better embeddings for comparison?

If so, which downstream task would you recommend for comparing python code similarity? ~I guess it's code-to-code zero shot search?~ (Just realised there is no fine-tuning settings for zero shot search.)

Thanks in advance!

**Comment by guoday** [comment]
Sure. Fine-tuning  them can makes it produce better embeddings for comparison. If you want to compare whether python files from two repositories, I suggest you can collect the positive pairs <code1,code2> that come from the same repository to fine-tune the models. Otherwise, zero-shot setting may be better.

**Comment by guoday** [comment]
I don't think BigCloneBench clone detection work, since the dataset only have 10 functions and it's hard to generalize to other domain. For code-to-code search in UniXcoder, the dataset comes from POJ datasets and it's also hard to generalize to other domain. In fact, for clone detection or code-to-code search tasks, the problem is the lack of supervised datasets. So this is the reason why we only use zero-shot embedding.

**Comment by lazyhope** [comment]
For python code it is difficult to find a clone benchmark, but I did managed to find one here: https://huggingface.co/datasets/PoolC/1-fold-clone-detection-600k-5fold, and I used it to fine-tune a modified GraphCodeBERT model which works well on clone detection task. However it doesn't scale well when I need to compare all functions from two repositories, as the model takes code pairs as input and I need to run the model O(n^2) times. That's why I want to use UniXCoder as it produce better semantic embeddings, and I want to calculate code embeddings similarity for later use rather than using a trained binary classifier.

Currently UniXCoder creates pretty good python embeddings from my testing, but now that I have this large dataset of python code pairs labeled with 0/1(similarity or not), can I use it to fine-tune the UnixCoder so that the embeddings of semantically similar python code is closer in the vector space (and vice versa)?

Sorry I am still learning transformers so may ask too many questions, and thank you so much for you time!
